src/gurt/telemetry.c

```diff
@@ -2597,41 +2846,32 @@ d_tm_add_ephemeral_dir(struct d_tm_node_t **node, size_t size_bytes,
 	if (rc != 0)
 		D_GOTO(fail_shmem, rc);
 
-	/* Add a link to the new region */
-	rc = add_metric(ctx, &link_node, D_TM_LINK, NULL, NULL, path);
-	if (rc != 0) {
-		D_ERROR("can't set up the link node, " DF_RC "\n", DP_RC(rc));
-		D_GOTO(fail_tracking, rc);
-	}
-
-	/* track attached regions within the parent shmem */
-	parent_shmem = get_shmem_for_key(ctx, link_node->dtn_shmem_key);
-	if (parent_shmem == NULL) {
-		D_ERROR("failed to get parent shmem pointer\n");
-		D_GOTO(fail_link, rc = -DER_NO_SHMEM);
-	}
+	d_tm_unlock_shmem();
+	if (tm_shmem.multiple_writer_lock)
+		D_MUTEX_UNLOCK(&ctx->shmem_root->sh_multiple_writer_lock);
 
-	D_ASSERT(link_node->dtn_type == D_TM_LINK);
-	link_metric                 = conv_ptr(parent_shmem, link_node->dtn_metric);
-	link_metric->dtm_data.value = key;
+	rc = attach_path_segment(key, path);
+	if (rc != 0)
+		D_GOTO(fail_attach, rc);
 
-	rc = get_free_region_entry(parent_shmem, &region_entry);
+	rc = sync_attached_segment_uid(path, key);
 	if (rc != 0)
-		D_GOTO(fail_link, rc);
-	region_entry->rl_key = key;
-	region_entry->rl_link_node = link_node;
+		D_GOTO(fail_sync, rc);
 
 	if (node != NULL)
 		*node = new_node;
 
-	d_tm_unlock_shmem();
-	if (tm_shmem.multiple_writer_lock)
-		D_MUTEX_UNLOCK(&ctx->shmem_root->sh_multiple_writer_lock);
 	return 0;
 
-fail_link:
-	invalidate_link_node(parent_shmem, link_node);
-fail_tracking:
+fail_sync:
+	d_tm_del_ephemeral_dir(path);
+	goto fail_unlock; /* shmem will be closed/destroyed already */
+fail_attach:
+	/* dropped locks before calling attach; have to reacquire */
+	d_tm_lock_shmem();
+	if (tm_shmem.multiple_writer_lock)
+		D_MUTEX_LOCK(&ctx->shmem_root->sh_multiple_writer_lock);
```

Comment:
Need to be careful to always take these in the same order to avoid deadlocks. Above we take the multi-writer lock first. I wonder if we should combine these two locks inside `d_tm_lock_shmem` to ensure reliable behavior. Is there any point where we call one without the other?

Reply:
Di added this extra lock... I agree that there's probably no reason to make callers have to manage them separately.

Reply:
Moved the multiple writer lock management into the existing lock/unlock functions. Seems to work well, and definitely simplifies the code.